[GTK][WPE] Error rendering the background image of a div when zoom is applied

magomez opened the pull request:
#### 1d790df1be6e35533981c6cccaa5a3ab46f09080
<pre>
[GTK][WPE] Error rendering the background image of a div when zoom is applied
<a href="https://bugs.webkit.org/show_bug.cgi?id=274823">https://bugs.webkit.org/show_bug.cgi?id=274823</a>

Reviewed by Carlos Garcia Campos.

When using the original image as the pattern source, check whether we&apos;re sampling
out of the bounds of the image to set the repeat or clamp flags.

* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::drawPattern):

Canonical link: <a href="https://commits.webkit.org/279565@main">https://commits.webkit.org/279565@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c432c11a8a2bbd92e7e63cdd16022c70dfcd2e86

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53703 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33070 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6220 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56984 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4429 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56007 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40559 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4267 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43506 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2897 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55800 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31288 "Found 60 new test failures: accessibility/ax-object-destroyed-on-reload.html, accessibility/basic-focusability.html, accessibility/body-element-aria-hidden.html, accessibility/button-inside-label-ax-text.html, accessibility/changing-aria-hidden-with-display-none-parent.html, accessibility/checkbox-mixed-value.html, accessibility/checkbox-radio-element-rect.html, accessibility/datetime/input-time-label-value.html, accessibility/display-contents/aria-hidden.html, accessibility/display-contents/dynamically-added-children.html ... (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46435 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24641 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28113 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3754 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2584 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/49840 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3934 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58578 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28868 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3980 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50913 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30066 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46597 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50259 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30999 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7967 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29843 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -901,9 +901,21 @@ void GraphicsContextSkia::drawPattern(NativeImage& nativeImage, const FloatRect&
+            bool repeatY = imageSampledRect.y() < 0 || std::trunc(imageSampledRect.height()) > nativeImage.size().height();
+            paint.setShader(image->makeShader(repeatX ? SkTileMode::kRepeat : SkTileMode::kClamp, repeatY ? SkTileMode::kRepeat : SkTileMode::kClamp, samplingOptions, &shaderMatrix));
+        } else
+            paint.setShader(image->makeShader(SkTileMode::kRepeat, SkTileMode::kRepeat, samplingOptions, &shaderMatrix));
```
Maybe we can initialize repeat X/Y as true, set it to false if needed inside the if and keep a single call to setShader

magomez replied:
Sure! I didn't worry too much about that yet, as I wanted to check the test results first, but I'll handle that in the final version.

magomez committed:
[GTK][WPE] Error rendering the background image of a div when zoom is applied
https://bugs.webkit.org/show_bug.cgi?id=274823

Reviewed by Carlos Garcia Campos.

When using the original image as the pattern source, check whether we're sampling
out of the bounds of the image to set the repeat or clamp flags.

* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::drawPattern):

Canonical link: https://commits.webkit.org/279565@main
